api/libproject/devstart.rb

```diff
@@ -751,9 +751,21 @@ def circle_deploy(cmd_name, args)
     end
   end
 
-  version = ENV.fetch("CIRCLE_TAG", "circle-ci-test")
-  promote = ENV["CIRCLE_BRANCH"] == "master" ? "--promote" : "--no-promote"
-  deploy(cmd_name, args + %W{--version #{version} #{promote}})
+  default_version = "circle-ci-test"
+  promote = "--no-promote"
+  if ENV["CIRCLE_BRANCH"] == "master"
+    # Note that --promote will generally be a no-op, as we expect
+    # circle-ci-master to always be serving 100% traffic. Pushing to an existing
+    # live version will immediately make those changes live. In the event that
+    # someone mistakenly pushes a different version manually, this --promote
+    # will restore us to the expected circle-ci-master version on the next
+    # commit.
+    promote = "--promote"
+    default_version = "circle-ci-master"
+  end
```

Comment:
It's not clear to me from the Circle docs if CIRCLE_BRANCH  and CIRCLE_TAG can both be set at the same time (say, when pushing a tagged commit to master) or whether one takes precedence or two builds trigger. We might just assert inside the `if` above that CIRCLE_TAG is not set...or else we could get surprising behavior below.

Reply:
Good point - I'd forgotten the exact semantics of tags. Reading the docs, I think these env vars should be mutually exclusive - I've added that assertion above.